dataplane/snappi/sai_dataplane.py

```diff
@@ -0,0 +1,334 @@
+import logging
+import time
+
+import dpkt
+from saichallenger.common.sai_dataplane import SaiDataplane
+from snappi import snappi
+
+BASE_TENGINE_PORT = 5555
+
+class SaiDataplaneImpl(SaiDataplane):
+
+    def __init__(self, exec_params):
+        self.alias = exec_params['alias']
+        self.mode = exec_params['mode']
+        super().__init__(exec_params)
+        self.flows = []
+
+    def init(self):
+        # Configure a new API instance where the location points to controller
+        # Ixia-C:       location = "https://<tgen-ip>:<port>"
+        # IxNetwork:    location = "https://<tgen-ip>:<port>", ext="ixnetwork"
+        # TRex:         location =         "<tgen-ip>:<port>", ext="trex"
+        if self.mode == 'ixnet':
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False, ext='ixnetwork')
+        elif self.mode == 'trex':
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False, ext='trex')
+        else:
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False)
+        logging.info("%s Starting connection to controller... " % time.strftime("%s"))
+
+        # Create an empty configuration to be pushed to controller
+        self.configuration = self.api.config()
+
+        # Configure two ports where the location points to the port location:
+        # Ixia-C:       port.location = "localhost:5555"
+        # IxNetwork:    port.location = "<chassisip>;card;port"
+        # TRex:         port.location = "localhost:5555"
+        for pid, port in enumerate(self.exec_params['port_groups']):
+            location = port.get('location', f"localhost:{BASE_TENGINE_PORT+pid}")
+            self.configuration.ports.port(name=port[port['init']], location=location)
+
+        cap = self.configuration.captures.capture(name="c1")[-1]
```

Comment:
would suggest adding new ports/captures using newly added `add` construct.

https://github.com/open-traffic-generator/conformance/blob/main/features/flows/headers/udp/test_udp_header.py#L42

https://github.com/open-traffic-generator/conformance/blob/main/features/flows/headers/udp/test_udp_header.py#L49

Reply:
So we will be able to capture per port? That's will be extremely helpful! Thanks!

Reply:
yes, indeed, you should be able to capture on per port basis.

Reply:
Added to backlog.